tests/unit/onyxTest.js

```diff
@@ -145,18 +147,88 @@ describe('Onyx', () => {
     it('should merge 2 arrays when it has no initial stored value for test key', () => {
         let testKeyValue;
         connectionID = Onyx.connect({
-            key: TEST_KEY,
+            key: ONYX_KEYS.TEST_KEY,
             initWithStoredValues: false,
             callback: (value) => {
                 testKeyValue = value;
             },
         });
 
-        Onyx.merge(TEST_KEY, ['test1']);
-        Onyx.merge(TEST_KEY, ['test2']);
+        Onyx.merge(ONYX_KEYS.TEST_KEY, ['test1']);
+        Onyx.merge(ONYX_KEYS.TEST_KEY, ['test2']);
         return waitForPromisesToResolve()
             .then(() => {
                 expect(testKeyValue).toEqual(['test1', 'test2']);
             });
     });
+
+    it('should properly set and merge when using mergeCollection', () => {
+        const valuesReceived = {};
+        let numberOfCallbacks = 0;
+
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.COLLECTION.TEST_KEY,
+            initWithStoredValues: false,
+            callback: (data) => {
+                ++numberOfCallbacks;
+                valuesReceived[data.ID] = data.value;
+            },
+        });
+
+        // The first time we call mergeCollection we'll be doing a multiSet internally
+        return Onyx.mergeCollection(ONYX_KEYS.COLLECTION.TEST_KEY, {
+            test_1: {
+                ID: 123,
+                value: 'one'
+            },
+            test_2: {
+                ID: 234,
+                value: 'two'
+            },
+            test_3: {
+                ID: 345,
+                value: 'three'
+            }
+        })
+            .then(() => {
+                // 2 key values to update and 2 new keys to add.
+                // MergeCollection will perform a mix of multiSet and multiMerge
+                Onyx.mergeCollection(ONYX_KEYS.COLLECTION.TEST_KEY, {
```

Comment:
Might be able to return this back to the `.then()` instead of returning `waitForPromisesToResolve()`?

Reply:
Tried it. Doesn't work when I return this since it fails here `expect(numberOfCallbacks).toEqual(7);` thinking the value is 3. I think it takes time for the callback to finish so `waitForPromisesToResolve` is maybe waiting for the callback to also finish which is why I get the correct values in expect when I use `waitForPromisesToResolve`

Reply:
>maybe waiting for the callback to also finish which is why I get the correct values

There is no async work happening in `keysChanged()` or the callbacks so I think it's something else...
I did notice we are returning the `Promise` created by `AsyncStorage.getAllKeys()` but not returning the inner promise here...

https://github.com/Expensify/react-native-onyx/blob/727c1347743ed35aac30393dca8bb530b15947bd/lib/Onyx.js#L512-L535